Nautilus/Documentation/tutorials/vehicle-module.md

```diff
@@ -0,0 +1,137 @@
+
+```csharp
+prefab.SetRecipe(new Crafting.RecipeData()
+{
+    craftAmount = 1,
+    Ingredients = new List<CraftData.Ingredient>()
+    {
+        new CraftData.Ingredient(TechType.HullReinforcementModule3),
+        new CraftData.Ingredient(TechType.AluminumOxyde, 2),
+        new CraftData.Ingredient(TechType.AdvancedWiringKit, 3)
+    }
+})
+    .WithFabricatorType(CraftTree.Type.SeamothUpgrades)
+    .WithStepsToFabricator("SeamothModules")
+    .WithCraftingTime(5f);
+```
+
+And finally, let's make the part that interest us: Adding the Upgrade Module Gadget.
+
+```csharp
+// This first function defines the equipment type and the quick slot type.
+prefab.SetUpgradeModule(EquipmentType.SeamothModule, QuickSlotType.Passive)
+    .WithDepthUpgrade(1700f, true)  // 1700 is the depth in meters. The boolean after defines if we want the value to be absolute or "relative", added to the default depth. 
+                                    // If it is on false, it will set the new crush depth to 1900 meters because the default depth of the Seamoth is 200 meters.
+                                    // Otherwise, it will set the new crush depth on 1700 meters.
+    .WithOnModuleAdded((Vehicle vehicleInstance, int slotId) =>
+    {
+        Subtitles.Add("Well done! The module is working! The new depth is 1700 meters.");
+    })
+    .WithOnModuleRemoved((Vehicle vehicleInstance, int slotId) => {
+        Subtitles.Add("Ah... You removed the upgrade. Take care of your hull.");
+    });
+```
+> [!NOTE]
+> You can do basically everything you want in these delegates, such as trigger story events, destroy the vehicle, add components to the vehicle instance, etc...
+
+## Chargeable self-destruct module
+
+Admitting we've already done the prefab info and the custom prefab, let's directly work only with the upgrade module gadget.
+
+```csharp
+var maxcharge = 50f;
+var cooldown = 10f;
+prefab.SetUpgradeModule(EquipmentType.VehicleModule, QuickSlotType.SelectableChargeable)
+    .WithMaxCharge(maxcharge)
+    .WithCooldown(cooldown)
+    .WithOnModuleAdded((Vehicle inst, int slotId) =>
+    {
+        Subtitles.Add("Self-destruct module installed. The module needs to be charged fully to detonate.");
+    })
+    .WithOnModuleRemoved((Vehicle inst, int slotId) =>
+    {
+        Subttiles.Add("Self-destruct module uninstalled.");
+    })
+    .WithOnModuleUsed((Vehicle inst, int slotID, float charge, float chargeScalar) =>
+    {
+        if(charge != maxCharge)
+        {
+            Subtitles.Add("Self-destruction sequence disengaged.")
+            return;
+        }
+        else
+        {
+            Subtitles.Add("Self-destruction sequence engaged.")
+            EngageSelfDestruct(inst, countdown);
+        }
+    });
+
+static void EngageSelfDestruct(Vehicle instance, float countdown)
+{
+    while(Time.time < (Time.time + countdown))
+    {
+        continue;
+    }
+    instance.liveMixin.Kill(DamageType.Explosive);
+}
```

Comment:
Funny example lol, but it seems like it would crash the game. Should be a coroutine, called like `UWE.CoroutineHost.StartCoroutine(EngageSelfDestruct)`, and instead of `continue`, put `yield return null`.

Reply:
So in an IEnumerable ?

Reply:
Ah no, as a CoroutineTask

Reply:
i forgor 💀

Reply:
Nope, the IEnumerator return type is what Unity uses for [coroutines](https://docs.unity3d.com/Manual/Coroutines.html).

Reply:
Btw, the problem before was that you were waiting for Time.time to change, but it never has a chance to change because time can only continue AFTER that method reaches its end.

Reply:
then should be Unity's StartCoroutine ?

Reply:
> Btw, the problem before was that you were waiting for Time.time to change, but it never has a chance to change because time can only continue AFTER that method reaches its end.

Fixed

Reply:
`UWE.CoroutineHost.StartCoroutine(...)` is better here because it can be called from anywhere

Reply:
ok alright, so the type of the function must be a CoroutineTask or a CoroutineHost ?

Reply:
Neither, IEnumerator should be the return type because Unity expects a coroutine to have that